cmds/init/init.go

```diff
@@ -169,46 +164,88 @@ func main() {
 		"/buildbin/rush",
 	}
 
-	noCmdFound := true
+	var cmdCount int
+	// systemd is "special". If we are supposed to run systemd, we're
+	// going to exec, and if we're going to exec, we're done here.
+	// systemd uber alles.
+	initFlags := cmdline.GetInitFlagMap()
+	// systemd gets upset when it discovers it isn't really process 1, so
+	// we can't start it in its own namespace. I just love systemd.
+	systemd, present := initFlags["systemd"]
+	systemdEnabled, boolErr := strconv.ParseBool(systemd)
+	if present && boolErr == nil && systemdEnabled == true {
+		v := cmdList[0]
+		debug("Exec %v", v)
+		if err := syscall.Exec(v, []string{v}, envs); err != nil {
+			log.Printf("Lucky you, systemd failed: %v", err)
+		}
+		// well, what a shame.
+		cmdList = cmdList[1:]
+		cmdCount++
+	}
+	// If the systemd failed for some reason, we might as well do the other things.
+
 	for _, v := range cmdList {
-		if _, err := os.Stat(v); !os.IsNotExist(err) {
-
-			noCmdFound = false
-
-			initFlags := cmdline.GetInitFlagMap()
-			// systemd gets upset when it discovers it isn't really process 1
-			systemd, present := initFlags["systemd"]
-			systemdEnabled, boolErr := strconv.ParseBool(systemd)
-			if !present || boolErr != nil || systemdEnabled == false {
-				cmd = exec.Command(v)
-				cmd.Env = envs
-				cmd.Stdin = os.Stdin
-				cmd.Stderr = os.Stderr
-				cmd.Stdout = os.Stdout
-				if *test {
-					cmd.SysProcAttr = &syscall.SysProcAttr{Cloneflags: cloneFlags}
-				} else {
-					cmd.SysProcAttr = &syscall.SysProcAttr{Setctty: true, Setsid: true, Cloneflags: cloneFlags}
-				}
-				debug("Run %v", cmd)
-				if err := cmd.Run(); err != nil {
-					log.Print(err)
-				}
+		if _, err := os.Stat(v); os.IsNotExist(err) {
+			continue
+		}
+
+		// inito is (optionally) created by the u-root command when
+		// the u-root initramfs is merged with an existing initramfs that has
+		// a /init. The name inito means "original /init"
+		// There may be an inito if we are building on
+		// an existing initramfs. All initos need their
+		// own pid space.
+		var cloneFlags uintptr
+		if v == "/inito" {
+			cloneFlags = uintptr(syscall.CLONE_NEWPID)
+		}
+
+		cmdCount++
+		cmd = exec.Command(v)
+		cmd.Env = envs
+		cmd.Stdin, cmd.Stdout, cmd.Stderr = os.Stdin, os.Stdout, os.Stderr
+		if *test {
+			cmd.SysProcAttr = &syscall.SysProcAttr{Cloneflags: cloneFlags}
+		} else {
+			cmd.SysProcAttr = &syscall.SysProcAttr{Setctty: true, Setsid: true, Cloneflags: cloneFlags}
+		}
+		debug("Run %v", cmd)
+		if err := cmd.Start(); err != nil {
+			log.Printf("Error starting %v: %v", v, err)
+			continue
+		}
+		for {
+			var s syscall.WaitStatus
+			var r syscall.Rusage
+			if p, err := syscall.Wait4(-1, &s, 0, &r); p == cmd.Process.Pid {
```

Comment:
I see their point. I don't know if it applies to us. SIGSTP eats wait status, so a SIGSTP handler might eat the exit status. how annoying. OTOH, we don't manage SIGSTP, this may not really apply.

Reply:
:+1: